code/graph_algorithms/src/centroid_decomposition/centroid_decompostition.cpp

```diff
@@ -0,0 +1,117 @@
+    std::vector<int> sizes_;
+    std::vector<bool> marked_;
+    Graph(int num) :
+        ver_(num),
+        adj_(num + 1),
+        centroidTree_(num + 1),
+        sizes_(num + 1),
+        marked_(num + 1)
+        {}
+
+    void addEdge(int x, int y, std::vector<std::vector<int>> &graph);
+    void calcSizes(int cur, int par, std::vector<std::vector<int>> graph);
+    int findCentroid(int cur, int par, int vertices, std::vector<std::vector<int>> graph);
+    void decomposeTree(int cur, int par, int total, int level, std::vector<std::vector<int>> graph);
+};
+
+void Graph::addEdge(int x, int y, std::vector<std::vector<int>> &graph)
+{
+    if (x < 0 || x > ver_)
+    {
+        throw std::out_of_range{"x is out of boundaries"};
+    }
+    if (y < 0 || y > ver_)
+    {
+        throw std::out_of_range{"y is out of boundaries"};
+    }
+    graph[x].push_back(y);
+    graph[y].push_back(x);
+}
+
+void Graph::calcSizes(int cur, int par, std::vector<std::vector<int>> graph)
+{
+    sizes_[cur] = 1;
+    for (const auto& to : graph[cur])
+    {
+        if (!(to == par || marked_[to] == true))
+        {
+            calcSizes (to, cur, graph);
+            sizes_[cur] += sizes_[to];
+        }
+    }
+}
+
+int Graph::findCentroid(int cur, int par, int vertices, std::vector<std::vector<int>> graph){
+    for (const auto& to : graph[cur])
+    {
+        if (!(to == par || marked_[to] == true))
+        {
+            if (sizes_[to] > vertices / 2)
+            {
+                return findCentroid(to, cur, vertices, graph);
+            }
+        }
+        return cur;
+    }
+}
+
+void Graph::decomposeTree(int cur, int par, int total, int level, std::vector<std::vector<int>> graph)
+{
+    calcSizes(cur, -1, graph);
+    int centroid = findCentroid(cur, -1, sizes_[cur], graph);
+    Level[level].push_back(centroid);
+    calcSizes(centroid, -1, graph);
+    marked_[centroid] = true;
+    for (const auto& to : adj_[centroid])
+    {
+        if(!(to == par || marked_[to] == true))
+        {
+            decomposeTree(to, cur, sizes_[to], level + 1, graph);
+            addEdge(cur, to, centroidTree_);
+        }
+    }
+}
+
+void showOutput()
+{
+    for (int i = 1; Level[i].empty() == false; ++i)
+    {
+        std::cout << "Centroids of level " << i << ": ";
+        bool more_than_1 = false;
+        for(const auto& j : Level[i])
+        {
+            if (more_than_1 == true)
+            {
+                std::cout << ", ";
+            }
+            std::cout << j;
+            more_than_1 = true;
+        }
+        std::cout << '\n';
+    }
+}
+
+int main(){
+    Graph tree(7);
+    Level.resize(8);
+    tree.addEdge(1, 2, tree.adj_);
+    tree.addEdge(1, 3, tree.adj_);
+    tree.addEdge(2, 4, tree.adj_);
+    tree.addEdge(4, 5, tree.adj_);
+    tree.addEdge(4, 6, tree.adj_);
+    tree.addEdge(2, 7, tree.adj_);
+    tree.decomposeTree(1, -1, tree.ver_, 1, tree.adj_);
+    showOutput();
```

Comment:
For `showOutput`, I would pass the Graph as an argument to it, and then iterate over it inside that function. Don't use the global variable `Level`, or keep track of it inside your graph class.

Reply:
@arnavb I didn't quite understand what you mean by this. I add centroids in certain "levels" with help of decomposeTree. If I understood you correctly, then I have to decompose the graph again, second time to write centroids into Level[i].